extensions/amp-analytics/0.1/events.js

```diff
@@ -447,24 +722,128 @@ export class TimerEventTracker extends EventTracker {
         'Bad maxTimerLength specification');
     const callImmediate = 'immediate' in timerSpec ?
         Boolean(timerSpec['immediate']) : true;
+    const timerStart = 'startSpec' in timerSpec ? timerSpec['startSpec'] : null;
+    user().assert(!timerStart || typeof timerStart == 'object',
+        'Bad timer start specification');
+    const timerStop = 'stopSpec' in timerSpec ? timerSpec['stopSpec'] : null;
+    user().assert((!timerStart && !timerStop) || typeof timerStop == 'object',
+        'Bad timer stop specification');
+
+    const timerId = this.generateTimerId_();
+    const isUnstoppableTimer = !timerStop;
+    let startBuilder;
+    let stopBuilder;
+    if (!!timerStart) {
+      const startTracker = this.getTracker(timerStart);
+      user().assert(startTracker, 'Cannot track timer start');
+      startBuilder = startTracker.add.bind(startTracker, context,
+          timerStart['on'], timerStart,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
+    }
+    if (!isUnstoppableTimer) {
+      const stopTracker = this.getTracker(timerStop);
+      user().assert(stopTracker, 'Cannot tracker timer stop');
+      stopBuilder = stopTracker.add.bind(stopTracker, context,
+          timerStop['on'], timerStop,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
+    }
 
-    const win = this.root.ampdoc.win;
-    const intervalId = win.setInterval(() => {
-      listener(this.createEvent_(eventType));
-    }, interval * 1000);
-    this.trackers_.push(intervalId);
-    win.setTimeout(() => {
-      this.removeTracker_(intervalId);
-    }, maxTimerLength * 1000);
-    if (callImmediate) {
-      listener(this.createEvent_(eventType));
+    const timerHandler = new TimerEventHandler(interval, maxTimerLength,
+        isUnstoppableTimer, callImmediate, startBuilder, stopBuilder);
+    this.trackers_[timerId] = timerHandler;
+
+    if (!timerStart) {
```

Comment:
Trying to understand the two classes.
`TimerEventTracker` handles `setInterval` and call `listener`
While `TimerEventHandler` handles when to ask `timerEventTracker` to start working. 
Can we make the `timerEventHandler` general enough, that it can handle every timer events' start and stop point. For example, even without `timerStart`, I can rely on `timerEventHandler` to get the correct point to start timer. 


Can we let the `TimerEventHandler` handle this during initiation instead? inside `TimerEventHandler` we can replace the check to `!timerStart` with `!startBuilder` ? Would that work. Then we just need to Can make at least the TimerEventTracker cleaner.

Reply:
Done.